dart/common/Factory.hpp

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2017, Graphics Lab, Georgia Tech Research Corporation
+ * Copyright (c) 2017, Personal Robotics Lab, Carnegie Mellon University
+ * All rights reserved.
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_FACTORY_HPP_
+#define DART_COMMON_FACTORY_HPP_
+
+#include <functional>
+#include <unordered_map>
+#include <memory>
+
+#include "dart/common/StlHelpers.hpp"
+
+namespace dart {
+namespace common {
+
+/// Implementation of the Abstract Factory Pattern.
+///
+/// Factory class is a pure static class (i.e., no need to create an instance to
+/// use this class).
+///
+/// Example:
+/// \code
+/// using CdFactory = Factory<std::string, CollisionDetector>;
+///
+/// CdFactory::registerCreator<FclCollisionDetector>("fcl");
+/// auto fclCd = CdFactory::create("fcl");
+/// \endcode
+template <typename KeyT,
+          typename BaseT,
+          template<typename...> class SmartPointerT = std::shared_ptr,
+          typename... Args>
+class Factory final
+{
+public:
+  using This = Factory<KeyT, BaseT, SmartPointerT>;
+  using CreatorReturnType = SmartPointerT<BaseT>;
+  using Creator = std::function<CreatorReturnType(Args...)>;
+  using CreatorMap = std::unordered_map<KeyT, Creator>;
+  using RegisterResult = std::pair<typename CreatorMap::iterator, bool>;
+
+  /// Registers a object creator function with a key.
+  static RegisterResult registerCreator(const KeyT& key, Creator creator);
+
+  /// Registers the default object creator function with a key.
+  template <typename Derived>
+  static RegisterResult registerCreator(const KeyT& key);
```

Comment:
This is a legacy code that was used to support `ScopedFactoryRegistrar`, which is removed. Will remove the return type as well.

> As an added note - why does the first call to registerCreator win? I would expect a later call to win, i.e. use operator[] instead of insert() to perform the insert operation.

:+1: